### DISCUSSION TITLE
Question: How "final" is the WASM API?

### DISCUSSION THREAD

**Discussion by asibahi** [harfbuzz/harfbuzz#4799] [Ideas]
Hello everyone,

While trying to implement the ways shaper API in `rustybuzz`, I inquired a lot on the Rust discord and reddit about the intricacies and idioms of wasm. One recurring theme was that the currently specified harfbuzz api falls short in a few ways.

First and most important: Blobs.
 
There is no consistent, safe way to copy binary blobs to the module from the host. WAMR (the library used currently by harfbuzz, provides `wasm_runtime_module_dup_data`, which seems to be a [custom, non-standards compliant, implementation](https://github.com/bytecodealliance/wasmtime/issues/8912). (While *not* searching for it, I finally found the [technical explanation behind it](https://bytecodealliance.github.io/wamr.dev/blog/understand-the-wamr-heaps/), not that it is useful for what I was trying to do as neither `wasmtime` nor `wasmi` expose such a thing).

The way I resorted to doing it, to make sure I am allocating in free memory, is to `grow` the memory and allocate there for every blob requested. For `table`s and `info` and `pos` this is not really a problem, but it becomes a problem when sending a lot of items repeatedly, such as glyph outlines, because every single glyph is an api call, so you end up allocating a lot of pages for a small amount of unneeded, wasteful memory. 

There are two solutions to this problem: the idiomatic WASM solution is `ExternRef`. While they're usable just fine from the host side, Rust from the module side doesn't have these types natively. There are crates to [handle](https://crates.io/crates/externref) it, but I have not tested yet to see how they work.

The *usual* solution for Rust (and, apparently, C) codebases is for the *module* to export two functions, `malloc` and `free`, with their usual trappings. Whenever the host needs to sends a blob to the module, it calls the exported `malloc` function with the required size (and alignment?), which returns a `I32` representing a pointer, write your data to said pointer, then call `free`

For example, these would be the wasm module's exports (I copied this code from one of the discord conversations I had:

```wat
(module
  ;; ...
  (export "malloc" (func $malloc))
  (export "free" (func $free))
  (export "someFunc" (func $someFunc))
  (export "memory" (memory $memory))
)
```

This would be the js code calling it 

```js
WebAssembly.instantiateStreaming(fetch("simple.wasm")).then((obj) =>
  const { malloc, free, someFunc, memory } = obj.instance.exports;
  const ptr = malloc(/* some size */);
  // write to ptr via memory
  someFunc(ptr);
  free(ptr);
);
```

I am not *quite* clear on how all this works, to be honest, but it is something I intend to experiment with. [This is an example from the Rust discord of how such API can be used from the host side,](https://github.com/Nemo157/docs.rs/blob/eb98c25eb9cc168dea5bffe89a64b353eb6b6ede/src/web/highlight.rs#L59) albeit for copying in strings.

=

Secondly, specifying the import environment and memory.

Perhaps a minor point, but the docs should specify the names for the  importing environment and memory exports. Currently the API uses the default LLVM enivronment settings of `"memory"` for memory and `"env"` for environment. The module writer does not need to know about those (as they're compiler defaults), but for implementing the API I had to and the only way of finding them was inspecting the `was` files of the compiled, existing, examples.

Customizing the environment name is fairly easy from the module side in Rust (simply add `#[link(wasm_import_module = "foo")]` to the list of `extern` imports.)  In [typst docs](https://typst.app/docs/reference/foundations/plugin/), you can see how they're specifying the environment with `typst_env`.  While their API is fairly simple compared to the extensive API defined for hardfuzz, it is a lot more clearly specified, I think.

This has the side benefit of allowing versioning API. The host (harfbuzz or rustybuzz) can specify different sets of API, with different environment names, and modules pick which ones to link against using that environment name, or even mix and match between them if wanted.

I do not know how to change that `memory` export name from within Rust, or if it is possible. Perhaps it is the most reasonable here to stick with the default until the Rust compiler catches up with the wasm spec (and then we can use `ExternRef` too)

(This whole point also directly accesses [this issue](https://github.com/harfbuzz/harfbuzz/issues/4457).)

As an aside, perhaps it would be better to define the API in terms of `wat` functions, like Typst does, instead of `C` functions.

=

Finally, could the API be refocused/redesigned in general?

The API could be expanded to cover new use cases. But from the examples only a small subset of the currently defined APIs are necessary. You perhaps do not need all the existing API, but only a subset, and even most of this subset is (as far as I can tell) unnecessary. I am not sure why harfbuzz wants a buffer token at all (is it possible to have multiple buffers? this wasn't clear to me at all).

`shaoe_with` could perhaps be made to work, not on *the* buffer, but on substrings of it, to use different features for different sections. The different `Face` tokens, which seemed entirely unnecessary to me, could be created by saving a map/dictionary for specific variation and feature settings. (So sth like `create_face_with_settings(variations, features)`). As I am unfamiliar with harfbuzz internals, I am not quite sure what is and is not needed. Y'all would have more insight on the subject.

=

Sorry for the long rant. It is a lot of things for what I think is clearly an interesting project. I have a mind to experiment with ways of doing different API (from Rust, backed by either rustybuzz or harfbuzz-rs) and see what makes sense and what does not. Maybe if I do that I can share it here and get y'al's input as well.

(should this have been an issue?)

**Comment by behdad** [comment]
Thanks Abdul. I'm currently traveling. Will reply properly in a few days. For now: the API is all but final. It's just a technology preview.

We also need to prefix the API: https://github.com/harfbuzz/harfbuzz/issues/4457
